### PR TITLE
[infra] Config file and some initial math formula implementation

### PR DESCRIPTION
**Pull request by Razvan00Rusu**
Implemented a JSON config file which probably isn't the best way to do it but should at least work for MVP. 

Also implemented some of the initial formulas done by the maths team, namely:

- Starting Variables
- Cower HP addition
- Initial Monster Resilience
- Initial Monster Damage

Something that came up is that we should be storing the number of agents that start each round somewhere, as this is used in some maths formula. As far as I can tell we aren't.. (sounds like a good use for a GitHub issue 🤔 ) :)

Still TODO: as I'm writing this PR I've realised I've forgotten to give each Agent a new unique ID since the initialisation... whoops will add a new commit to fix this.

### CODE REVIEW

File: pkg/infra/config/config.json [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+  "attackStrengthStart": 1000,
+  "shieldStrengthStart": 1000,
+  "thresholdPercentage": 0.6,
+  "agentConfig": [
+    {
+      "strategy": "random",
+      "quantity": 100
+    }
+  ]
+}
```
Comment (neeldug) [inline pkg/infra/config/config.json:14]:
```suggestion
```

File: pkg/infra/main.go [hunk 2]
```diff
@@ -16,25 +19,16 @@ Whereby the peer channels enable bidirectional p2p communication
...
 func main() {
-	agentMap, stateChannels, decisionChannels, globalState := initialise()
-	gameLoop(globalState, agentMap, decisionChannels, stateChannels)
+	agentMap, stateChannels, decisionChannels, globalState, gameConfig := initialise()
+	gameLoop(globalState, agentMap, decisionChannels, stateChannels, gameConfig)
 }
 
-func gameLoop(globalState state.State, agentMap map[uint]agent.Agent, decisionChannels map[uint]<-chan decision.Decision, stateChannels map[uint]chan<- state.State) {
-	for globalState.CurrentLevel = 0; globalState.CurrentLevel < numLevels; globalState.CurrentLevel++ {
+func gameLoop(globalState state.State, agentMap map[uint]agent.Agent, decisionChannels map[uint]<-chan decision.Decision, stateChannels map[uint]chan<- state.State, gameConfig config.GameConfig) {
```
Comment (neeldug) [inline pkg/infra/main.go:27]: These functions are starting to take a lot of parameters... We should probably start sticking them into structs to tidy things up a bit.

File: pkg/infra/main.go [hunk 3]
```diff
@@ -70,21 +66,58 @@ func gameLoop(globalState state.State, agentMap map[uint]agent.Agent, decisionCh
 			agentState.BonusDefense = shieldLoot[allocatedShield]
 			weaponLoot, _ = commons.DeleteElFromSlice(weaponLoot, allocatedWeapon)
 			shieldLoot, _ = commons.DeleteElFromSlice(shieldLoot, allocatedShield)
-
 		}
 	}
 }
 
-func initialise() (map[uint]agent.Agent, map[uint]chan<- state.State, map[uint]<-chan decision.Decision, state.State) {
+func initialise() (map[uint]agent.Agent, map[uint]chan<- state.State, map[uint]<-chan decision.Decision, state.State, config.GameConfig) {
```
Comment (neeldug) [inline pkg/infra/main.go:73]: > These functions are starting to take a lot of parameters... We should probably start sticking them into structs to tidy things up a bit.

File: pkg/infra/main.go [hunk 4]
```diff
@@ -70,21 +66,58 @@ func gameLoop(globalState state.State, agentMap map[uint]agent.Agent, decisionCh
...
+	}
+	return agentMap, stateChannels, decisionChannels, globalState, gameConfig
+}
+
+func instantiateAgent[S agent.Strategy](gameConfig config.GameConfig,
+	stateChannels map[uint]chan<- state.State,
+	decisionChannels map[uint]<-chan decision.Decision,
+	agentMap map[uint]agent.Agent,
+	agentStateMap map[uint]state.AgentState,
+	strategy S) {
```
Comment (neeldug) [inline pkg/infra/main.go:118]: My bad about this one:

> These functions are starting to take a lot of parameters... We should probably start sticking them into structs to tidy things up a bit.

**Comment by BusterBlack** [comment]
the equations requiring the number of agents refers purely to the starting number of agents... otherwise, in the case of the monsters, their resilience and damage would reduce as the number of agents reduces.

**Comment by Razvan00Rusu** [comment]
> the equations requiring the number of agents refers purely to the starting number of agents... otherwise, in the case of the monsters, their resilience and damage would reduce as the number of agents reduces. 
> 
> 

Oh cool. I'm not sure why I thought that wouldn't be the case tbh, I'll remove the comment

**Pull request merged**